File: src/features/vault/api/lock.ts
// App unlock, backed by the Android Keystore via react-native-keychain.
//
// Not a cosmetic prompt over the UI: there is a secret in hardware-backed
// storage whose access policy requires the device owner to authenticate, so
// getting into the app means satisfying the Keystore — the same mechanism that
// will guard the E2E message keys.
//
// react-native-keychain is already in the dev-client APK and contributes
// USE_BIOMETRIC to the merged manifest, so this needs no native rebuild.
//
// IMPORTANT: creating a key that requires user authentication on a phone with
// nothing enrolled makes Android launch its fingerprint-enrollment flow. That
// must never happen as a side effect of opening the app, so the key is only
// ever created from an explicit opt-in — see `enableLock`.
import * as Keychain from 'react-native-keychain';

const SERVICE = 'app.echo.unlock';
const ACCOUNT = 'echo';

/** The value is irrelevant; reaching it is the proof. */
const TOKEN = 'unlocked';

export type LockOutcome =
  | { ok: true }
  | { ok: false; reason: 'cancelled' | 'unavailable' | 'error'; message?: string };

const guarded = {
  service: SERVICE,
  // Biometrics first, device PIN as fallback. Biometry-only would shut out a
  // phone whose owner uses a PIN.
  accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_ANY_OR_DEVICE_PASSCODE,
  accessible: Keychain.ACCESSIBLE.WHEN_UNLOCKED_THIS_DEVICE_ONLY,
} as const;

const prompt = {
  title: 'Unlock Echo',
  subtitle: 'Your messages, keys and wallet are on this phone',
  cancel: 'Cancel',
};

const looksCancelled = (e: unknown) => {
  const text = String(e).toLowerCase();
  return text.includes('cancel') || text.includes('code: 13');
};

/** Does this phone have biometric hardware at all? Says nothing about whether
 *  anything is actually enrolled — Android has no such check outside
 *  expo-local-authentication's `isEnrolledAsync`, which needs a native rebuild. */
export async function hasBiometricHardware(): Promise<boolean> {
  try {
    return (await Keychain.getSupportedBiometryType()) !== null;
  } catch {
    return false;
  }
}

/** Has the user already turned the lock on? Does not authenticate or prompt. */
export async function isLockEnabled(): Promise<boolean> {
  try {
    return await Keychain.hasGenericPassword({ service: SERVICE });
  } catch {
    return false;
  }
}

/**
 * Turn the lock on. Only call this from a deliberate user action — on a phone
 * with no fingerprint enrolled, Android may open its enrollment flow here, and
 * that is only acceptable when someone has just asked for exactly this.
 */
export async function enableLock(): Promise<LockOutcome> {
  try {
    await Keychain.setGenericPassword(ACCOUNT, TOKEN, guarded);
    return { ok: true };
  } catch (e) {
    if (looksCancelled(e)) return { ok: false, reason: 'cancelled' };
    return { ok: false, reason: 'error', message: e instanceof Error ? e.message : String(e) };
  }
}

/** Ask the device owner to prove who they are. Only meaningful once enabled. */
export async function unlock(): Promise<LockOutcome> {
  try {
    const held = await Keychain.getGenericPassword({
      ...guarded,
      authenticationPrompt: prompt,
    });
    return held ? { ok: true } : { ok: false, reason: 'cancelled' };
  } catch (e) {
    if (looksCancelled(e)) return { ok: false, reason: 'cancelled' };
    return { ok: false, reason: 'error', message: e instanceof Error ? e.message : String(e) };
  }
}

/** Turn the lock off, or recover a demo phone that got into a bad state. */
export async function forgetLock(): Promise<void> {
  try {
    await Keychain.resetGenericPassword({ service: SERVICE });
  } catch {
    // Nothing stored — already where we wanted to be.
  }
}

File: flutter/lib/features/vault/lock.dart
// App unlock, backed by a hardware-backed keystore.
//
// Not a cosmetic prompt over the UI: there is a secret in hardware-backed
// storage whose access policy requires the device owner to authenticate, so
// getting into the app means satisfying the keystore — the same mechanism
// that will guard the E2E message keys.
//
// Port of src/features/vault/api/lock.ts (react-native-keychain,
// biometry-or-device-passcode) onto local_auth + flutter_secure_storage:
// [BiometricAppLock] drives the real BiometricPrompt, gated by an "enabled"
// flag persisted in the platform's secure storage so it survives restarts.
// This defines the contract [AppLock] the adapter satisfies; the headless
// fake used by tests (platform channels don't run under `flutter test`)
// lives in test/features/vault_fakes.dart.
//
// IMPORTANT: on a phone with nothing enrolled, creating a key that requires
// user authentication makes Android launch its fingerprint-enrollment flow.
// That must never happen as a side effect of opening the app — [enable] must
// only ever run from an explicit opt-in, never from [hasBiometricHardware] or
// [isEnabled].
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:local_auth/local_auth.dart';

enum LockFailureReason { cancelled, unavailable, error }

class LockOutcome {
  const LockOutcome.ok() : ok = true, reason = null, message = null;

  const LockOutcome.failure(this.reason, {this.message}) : ok = false;

  final bool ok;
  final LockFailureReason? reason;
  final String? message;
}

/// Contract a native hardware-backed lock must satisfy.
abstract class AppLock {
  /// Does this phone have biometric hardware at all? Says nothing about
  /// whether anything is actually enrolled.
  Future<bool> hasBiometricHardware();

  /// Has the user already turned the lock on? Does not authenticate or prompt.
  Future<bool> isEnabled();

  /// Turn the lock on. Only call this from a deliberate user action.
  Future<LockOutcome> enable();

  /// Ask the device owner to prove who they are. Only meaningful once enabled.
  Future<LockOutcome> unlock();

  /// Turn the lock off, or recover a demo phone that got into a bad state.
  Future<void> forget();
}

/// Real adapter: BiometryAny-or-device-passcode via local_auth, gated by an
/// "enabled" flag persisted in flutter_secure_storage (Android
/// Keystore-backed EncryptedSharedPreferences) so the lock stays on across
/// restarts. `biometricOnly: false` mirrors the upstream RN adapter's
/// `BIOMETRY_ANY_OR_DEVICE_PASSCODE`: biometrics first, device PIN/pattern as
/// fallback, so a phone whose owner uses a PIN isn't shut out.
class BiometricAppLock implements AppLock {
  BiometricAppLock({LocalAuthentication? auth, FlutterSecureStorage? storage})
    : _auth = auth ?? LocalAuthentication(),
      _storage = storage ?? const FlutterSecureStorage();

  static const _enabledKey = 'echo.vault.lock_enabled';

  final LocalAuthentication _auth;
  final FlutterSecureStorage _storage;

  @override
  Future<bool> hasBiometricHardware() async {
    try {
      return await _auth.canCheckBiometrics;
    } catch (_) {
      return false;
    }
  }

  @override
  Future<bool> isEnabled() async {
    try {
      return await _storage.read(key: _enabledKey) == 'true';
    } catch (_) {
      return false;
    }
  }

  @override
  Future<LockOutcome> enable() async {
    if (!await hasBiometricHardware()) {
      return const LockOutcome.failure(LockFailureReason.unavailable);
    }
    try {
      await _storage.write(key: _enabledKey, value: 'true');
      return const LockOutcome.ok();
    } catch (e) {
      return LockOutcome.failure(
        LockFailureReason.error,
        message: e.toString(),
      );
    }
  }

  @override
  Future<LockOutcome> unlock() async {
    if (!await isEnabled()) {
      return const LockOutcome.failure(LockFailureReason.unavailable);
    }
    try {
      final didAuthenticate = await _auth.authenticate(
        localizedReason: 'Unlock Echo',
      );
      return didAuthenticate
          ? const LockOutcome.ok()
          : const LockOutcome.failure(LockFailureReason.cancelled);
    } on LocalAuthException catch (e) {
      return LockOutcome.failure(
        _reasonFor(e.code),
        message: e.description ?? e.code.name,
      );
    } catch (e) {
      return LockOutcome.failure(
        LockFailureReason.error,
        message: e.toString(),
      );
    }
  }

  @override
  Future<void> forget() async {
    try {
      await _storage.delete(key: _enabledKey);
    } catch (_) {
      // Nothing stored — already where we wanted to be.
    }
  }

  static LockFailureReason _reasonFor(LocalAuthExceptionCode code) =>
      switch (code) {
        LocalAuthExceptionCode.userCanceled ||
        LocalAuthExceptionCode.systemCanceled ||
        LocalAuthExceptionCode.userRequestedFallback =>
          LockFailureReason.cancelled,
        LocalAuthExceptionCode.noCredentialsSet ||
        LocalAuthExceptionCode.noBiometricsEnrolled ||
        LocalAuthExceptionCode.noBiometricHardware ||
        LocalAuthExceptionCode.biometricHardwareTemporarilyUnavailable =>
          LockFailureReason.unavailable,
        _ => LockFailureReason.error,
      };
}
